yt/frontends/sph/data_structures.py

```diff
@@ -104,19 +129,25 @@ def _initialize_index(self):
 
     def _generate_kdtree(self, fname):
         from cykdtree import PyKDTree
-        if os.path.exists(fname):
-            mylog.info('Loading KDTree from %s' % os.path.basename(fname))
-            kdtree = PyKDTree.from_file(fname)
-            if kdtree.data_version != self.ds._file_hash:
-                mylog.info('Detected hash mismatch, regenerating KDTree')
-            else:
-                self._kdtree = kdtree
-                return
+        if fname is not None:
+            if os.path.exists(fname):
+                mylog.info('Loading KDTree from %s' % os.path.basename(fname))
+                kdtree = PyKDTree.from_file(fname)
+                if kdtree.data_version != self.ds._file_hash:
+                    mylog.info('Detected hash mismatch, regenerating KDTree')
+                else:
+                    self._kdtree = kdtree
+                    return
         positions = []
         for data_file in self.data_files:
-            for _, ppos in self.io._yield_coordinates(
+            if fname is None:
+                for _, ppos in self.io._yield_coordinates(
+                    data_file):
+                    positions.append(ppos)
+            else:
+                for _, ppos in self.io._yield_coordinates(
                     data_file, needed_ptype=self.ds._sph_ptype):
```

Comment:
modified this to work with the test sph data sets

Reply:
Rather than fixing this here I think it would be more appropriate to add a `needed_ptype` keyword argument to the `StreamDataset` implementation of `_yield_coordinates`.